store/v2/multi/snapshot.go

```diff
@@ -0,0 +1,259 @@
+		// zlib default compression level 6
+		// level 0 :  no compression and fastest
+		// ....
+		// level 7 : average compression and average speed
+		// ...
+		// level 9 : highest compression and speed is slower
+		zWriter, err := zlib.NewWriterLevel(bufWriter, 7)
+		if err != nil {
+			chunkWriter.CloseWithError(sdkerrors.Wrap(err, "zlib failure"))
+			return
+		}
+		defer func() {
+			if err := zWriter.Close(); err != nil {
+				chunkWriter.CloseWithError(err)
+			}
+		}()
+		protoWriter := protoio.NewDelimitedWriter(zWriter)
+		defer func() {
+			if err := protoWriter.Close(); err != nil {
+				chunkWriter.CloseWithError(err)
+			}
+		}()
+
+		var sKeys []string
+		// sending the snapshot store schema
+		for sKey := range vs.schema {
+			if vs.schema[sKey] == storetypes.StoreTypePersistent {
+				sKeys = append(sKeys, sKey)
+			}
+		}
+
+		sort.Slice(sKeys, func(i, j int) bool {
+			return strings.Compare(sKeys[i], sKeys[j]) == -1
+		})
+
+		var storeByteKeys [][]byte
+		for _, sKey := range sKeys {
+			storeByteKeys = append(storeByteKeys, []byte(sKey))
+		}
+
+		err = protoWriter.WriteMsg(&storetypes.SnapshotItem{
+			Item: &storetypes.SnapshotItem_Schema{
+				Schema: &storetypes.SnapshotSchema{
+					Keys: storeByteKeys,
+				},
+			},
+		})
+		if err != nil {
+			chunkWriter.CloseWithError(err)
+			return
+		}
+
+		for _, sKey := range sKeys {
+			subStore, err := vs.getSubstore(sKey)
+			if err != nil {
+				chunkWriter.CloseWithError(err)
+				return
+			}
+
+			err = protoWriter.WriteMsg(&storetypes.SnapshotItem{
+				Item: &storetypes.SnapshotItem_Store{
+					Store: &storetypes.SnapshotStoreItem{
+						Name: sKey,
+					},
+				},
+			})
+			if err != nil {
+				chunkWriter.CloseWithError(err)
+				return
+			}
+
+			iter := subStore.Iterator(nil, nil)
+			for ; iter.Valid(); iter.Next() {
+				err = protoWriter.WriteMsg(&storetypes.SnapshotItem{
+					Item: &storetypes.SnapshotItem_KV{
+						KV: &storetypes.SnapshotKVItem{
+							Key:   iter.Key(),
+							Value: iter.Value(),
+						},
+					},
+				})
+				if err != nil {
+					chunkWriter.CloseWithError(err)
+					return
+				}
+			}
+			iter.Close()
+		}
+	}()
+
+	return ch, nil
+}
+
+// Restore implements snapshottypes.Snapshotter.
+func (rs *Store) Restore(height uint64, format uint32, chunks <-chan io.ReadCloser, ready chan<- struct{}) error {
+	if format != snapshottypes.CurrentFormat {
+		return sdkerrors.Wrapf(snapshottypes.ErrUnknownFormat, "format %v", format)
+	}
+
+	if height == 0 {
+		return sdkerrors.Wrap(sdkerrors.ErrLogic, "cannot restore snapshot at height 0")
+	}
+	if height > uint64(math.MaxInt64) {
+		return sdkerrors.Wrapf(snapshottypes.ErrInvalidMetadata,
+			"snapshot height %v cannot exceed %v", height, int64(math.MaxInt64))
+	}
+
+	versions, err := rs.stateDB.Versions()
+	if err != nil {
+		return sdkerrors.Wrapf(err, "error while getting the snapshot versions at height %v", height)
+	}
```

Comment:
We are checking the state is empty or not before restoring the data here.

 We are restoring the snapshot version for particular block height here 
 ```go
 # It will create snapshot version at this height
 rs.commit(height)
 ```
 https://github.com/cosmos/cosmos-sdk/pull/10794/files#diff-38eb2b92590d839b88a4ef3431f90cb5ee8cf678025cc4f2361391a2f9a10296R226